vsphere/internal/virtualdevice/virtual_machine_device_subresource.go

```diff
@@ -547,57 +548,98 @@ func swapSCSIDevice(l object.VirtualDeviceList, device types.BaseVirtualSCSICont
 	return spec, nil
 }
 
-// NormalizeSCSIBus checks the SCSI controllers on the virtual machine and
+// NormalizeBus checks the storage controllers on the virtual machine and
 // either creates them if they don't exist, or migrates them to the specified
 // controller type. Devices are migrated to the new controller appropriately. A
 // spec slice is returned with the changes.
 //
 // The first number of slots specified by count are normalized by this
 // function. Any others are left unchanged.
-func NormalizeSCSIBus(l object.VirtualDeviceList, ct string, count int, st string) (object.VirtualDeviceList, []types.BaseVirtualDeviceConfigSpec, error) {
-	log.Printf("[DEBUG] NormalizeSCSIBus: Normalizing first %d controllers on SCSI bus to device type %s", count, ct)
+func NormalizeBus(l object.VirtualDeviceList, d *schema.ResourceData) (object.VirtualDeviceList, []types.BaseVirtualDeviceConfigSpec, error) {
+	scsiCount := d.Get("scsi_controller_count").(int)
+	scsiType := d.Get("scsi_type").(string)
+	scsiSharing := d.Get("scsi_bus_sharing").(string)
+	sataCount := d.Get("sata_controller_count").(int)
+	ideCount := d.Get("ide_controller_count").(int)
 	var spec []types.BaseVirtualDeviceConfigSpec
-	ctlrs := make([]types.BaseVirtualSCSIController, count)
+	scsiCtlrs := make([]types.BaseVirtualSCSIController, scsiCount)
+	sataCtlrs := make([]types.BaseVirtualSATAController, sataCount)
+	ideCtlrs := make([]*types.VirtualIDEController, ideCount)
 	// Don't worry about doing any fancy select stuff here, just go thru the
 	// VirtualDeviceList and populate the controllers.
-	iCount := int32(count)
+	log.Printf("[DEBUG] NormalizeBus: Normalizing first %d controllers on SCSI bus to device type %s", scsiCount, scsiType)
+	log.Printf("[DEBUG] NormalizeBus: Normalizing first %d controllers on SATA bus", sataCount)
+	log.Printf("[DEBUG] NormalizeBus: Normalizing first %d controllers on IDE bus", ideCount)
 	for _, dev := range l {
-		if sc, ok := dev.(types.BaseVirtualSCSIController); ok {
-			if busNumber := sc.GetVirtualSCSIController().BusNumber; busNumber < iCount {
-				ctlrs[busNumber] = sc
+		switch ctlr := dev.(type) {
+		case types.BaseVirtualSCSIController:
+			if busNumber := ctlr.GetVirtualSCSIController().BusNumber; busNumber < int32(scsiCount) {
+				scsiCtlrs[busNumber] = ctlr
+			}
+		case types.BaseVirtualSATAController:
+			if busNumber := ctlr.GetVirtualSATAController().BusNumber; busNumber < int32(sataCount) {
+				sataCtlrs[busNumber] = ctlr
+			}
+		case *types.VirtualIDEController:
+			if busNumber := ctlr.GetVirtualController().BusNumber; busNumber < int32(ideCount) {
+				ideCtlrs[busNumber] = ctlr
 			}
 		}
 	}
```

Comment:
Just wondering if it would make sense at this point to check if all three controller lists are full or not. Just a quick check that the device list contained all the devices we expected to find.

Reply:
It could help by allowing us to skip the next few loops, but it is ok to not have all the expected controllers found. In the next section of the function, any controllers that are `nil` are created. This is where new or missing controllers are found and created.